master/internal/api_workspace.go

```diff
@@ -996,12 +1061,16 @@ func (a *apiServer) DeleteWorkspace(
 		}
 
 		// Delete the auto-generated namespace (if it exists) and its resources in Kubernetes.
-		// TODO(saloni): Cleanup request processing workers if the auto-generated namespace is not
-		// bound to any other workspaces.
 		err = a.m.rm.DeleteNamespace(*namespaceName)
 		if err != nil {
 			return nil, err
 		}
+		for _, v := range deletedBindings {
```

Comment:
Would love some feedback about where to put this code. One option is putting it right after deleting the bindings from the db (after line 1049) and the other option is where I have put it in my PR. The caveat with the first option is that, incase the workspace deletion errors out, we would have removed these namespaces from our interfaces in the RM code - this is actually why we are deferring deleting the auto-created namespaces from k8 (function DeleteNamespace) to after the workspace has been deleted. With what we have currently, there is repeated code twice.

My thinking is that even if the workspace deletion fails, we are already removing the bindings from the database so, it should be okay to remove it from our Interfaces. But, I have taken the safer route in the PR.

Let me know your thoughts! cc: @amandavialva01

Reply:
Maybe I'm missing something, but it looks like the workspace deletion happens before the bindings get deleted? So in this case wouldn't putting the namespace deletion right after the delete query work fine? 

Otherwise, if we're deleting the workspace separately from the bindings, I'd expect the bindings to stay intact in the event of a workspace delete failure. It should ideally all be part of the transaction. Perhaps something like a cascade delete on the workspace table to workspace bindings?

Reply:
Sorry, just saw this! +1 to Eric's comment but also, I think we already have an on-delete cascade for the workspace-namespace bindings [ref here](https://github.com/determined-ai/determined/blob/wksp-namespace-binding/master/static/migrations/20240612140846_add-workspace-namespace-bindings-table.tx.up.sql#L2) , so I think it's better to handle the namespace deletion twice (as is in this PR) for each case. 
But taking a closer look at it actually, I think the best solution would be to just remove the `delete_workspace` query from the if-block because looking at the `deleteWorkspace` function, things are still handled cleanly if `len(projects)`is 0. 
So maybe instead of 

```
if len(projects) == 0 {
		err = a.m.db.QueryProto("delete_workspace", holder, req.Id)
		if err != nil {
			return nil, fmt.Errorf("error deleting workspace (%d): %w", req.Id, err)
		}

		// Delete the auto-generated namespace (if it exists) and its resources in Kubernetes.
		// TODO(saloni): Cleanup request processing workers if the auto-generated namespace is not
		// bound to any other workspaces.
		err = a.m.rm.DeleteNamespace(*namespaceName)
		if err != nil {
			return nil, err
		}

		return &apiv1.DeleteWorkspaceResponse{Completed: true}, nil
	}
	go func() {
		a.deleteWorkspace(ctx, req.Id, projects)
	}()
```

we can do 
```
go func() {
		a.deleteWorkspace(ctx, req.Id, projects)
	}()
	err = a.m.rm.DeleteNamespace(*namespaceName)
		if err != nil {
			return nil, err
		}
if len(projects) == 0 {
return &apiv1.DeleteWorkspaceResponse{Completed: true}, nil
}

```
Since bindings are deleted in the db on a delete cascade, this is all the cleanup we would need to do!

Reply:
ended up handling it slightly differently! Let me know what you think @amandavialva01  @eecsliu